# Regular Expressions

Regular expressions are powerful tools to extract *structured information* from *unstructured text.*  For example, suppose that we are scraping Twitter data, and we'd like to extract a list of all the mentions and hashtags in a tweet. Our raw data might look something like this: 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Our Great American Model was built on tough (very strong!!) parametric assumptions! <br><br>But FAR LEFT elitists living in coastal TANGENT SPACES (out of touch!) want to throw these out. Not on my watch!!<a href="https://twitter.com/hashtag/statstwitter?src=hash&amp;ref_src=twsrc%5Etfw">#statstwitter</a> <a href="https://twitter.com/hashtag/epitwitter?src=hash&amp;ref_src=twsrc%5Etfw">#epitwitter</a> <a href="https://twitter.com/hashtag/rstats?src=hash&amp;ref_src=twsrc%5Etfw">#rstats</a> <a href="https://twitter.com/hashtag/math?src=hash&amp;ref_src=twsrc%5Etfw">#math</a> <a href="https://twitter.com/hashtag/AI?src=hash&amp;ref_src=twsrc%5Etfw">#AI</a> <a href="https://twitter.com/hashtag/DataScience?src=hash&amp;ref_src=twsrc%5Etfw">#DataScience</a> <a href="https://twitter.com/hashtag/python?src=hash&amp;ref_src=twsrc%5Etfw">#python</a> <a href="https://twitter.com/hashtag/Science?src=hash&amp;ref_src=twsrc%5Etfw">#Science</a></p>&mdash; Statistician Trump (@StatisticianTr2) <a href="https://twitter.com/StatisticianTr2/status/1281959378371969024?ref_src=twsrc%5Etfw">July 11, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>    

We'd like to extract the hashtags from this tweet. For example, we'd like to write a function `collect_hashtags()` with the following output: 

```python
collect_hashtags(tw)
['statstwitter', 'epitwitter', 'rstats', 'math', 'AI', 'DataScience', 'python', 'Science']
```

We could then use this function on many tweets in order to conduct an analysis of what people are talking about on Twitter. How can we recognize the hashtags? 

If you're familiar with Twitter, you know that a hashtag consists of the symbol \#, followed by one or more letters, which may or may not be capitalized. A space `" "` terminates the hashtag. 

This is an informal description of a *pattern* -- a rule for detecting hashtags in text. In this case, the rule is: 

> Find a `#`. Then, continue through letters and numbers until a space `" "` is reached.

Regular expressions allow us to formally construct and use patterns to obtain structured data like hashtags from unstructured text. They are an extremely powerful tool in any applications in which we need to work with text data. 

To work with regular expressions, we need a few functions from the `re` package. 

In [1]:
import re

Here's a plaintext representation of our tweet. 

In [2]:
tw = "Our Great American Model was built on tough (very strong!!) parametric assumptions! But FAR LEFT elitists living in coastal TANGENT SPACES (out of touch!) want to throw these out. Not on my watch!! #statstwitter #epitwitter #rstats #math #AI #DataScience #python #Science" 
tw

'Our Great American Model was built on tough (very strong!!) parametric assumptions! But FAR LEFT elitists living in coastal TANGENT SPACES (out of touch!) want to throw these out. Not on my watch!! #statstwitter #epitwitter #rstats #math #AI #DataScience #python #Science'

The first thing we need to do is construct a *pattern* that matches the pieces of text that we want to find. Patterns are represented as *raw strings*, that is, they are preceded by `r` outside quotes. Raw strings don't process special characters. For example, the string `"\n"` has just one character (the special newline character), but the string `r"\n"` has two (`"\"` and `"n"`).  

In [4]:
len("\n"), len(r"\n")

(1, 2)

Now let's start pattern matching. Our main tool is the function `re.search()`. This function finds the very first match of the specified pattern. 

In [5]:
pattern = r"#"
result = re.search(pattern, tw)
result

<re.Match object; span=(198, 199), match='#'>

This says that the first match of the pattern `#` occurred at index 198. We can extract either the location of the match or the substring that produced the match. For the latter, we use the `group()` method -- we'll explain this name in a future lecture. 

In [6]:
result.span()

(198, 199)

In [8]:
result.group()

'#'

In this case, the matching substring is just a single character. Let's make things a bit more interesting -- we'll look for the first hashtag that begins with `"#epi"`:  

In [9]:
pattern = r"#epi"
result = re.search(pattern, tw)
result

<re.Match object; span=(212, 216), match='#epi'>

Let's check that the span corresponds to the location in the original string:  

In [10]:
sp = result.span()
tw[sp[0]:sp[1]]

'#epi'

We would have gotten the same result by checking `result.group()`: 

In [11]:
result.group()

'#epi'

## Various Syntax 

The regular expressions engine has a lot of syntax options that can help you easily express very complicated patterns. Here are a few of the most important ones. 

In [12]:
re.search(r"rk", "kirk")

<re.Match object; span=(2, 4), match='rk'>

In [13]:
# search always takes the FIRST match
re.search(r"rk", "kirk kirk")

<re.Match object; span=(2, 4), match='rk'>

In [15]:
# Repeated characters
# o* matches any segment of o's. May include no o's. 
# o+ matches any segment with at least one o

s = "Sisk Siskooooooo"

re.search(r"ko+", s)

<re.Match object; span=(8, 16), match='kooooooo'>

In [21]:
# Bracket expressions
# [A-z] matches any letter, upper or lower case
# try [A-Z] or [a-z]
# Add + to match continuous strings of letters (i.e. words)

s = "Siskoooooo in DS9"

re.search(r"[A-Z]+[0-9]", s)

<re.Match object; span=(14, 17), match='DS9'>

## Back to Twitter

Now we're ready to try finding hashtags. Each one is a `#` character followed by a string of letters and numbers, regardless of case, with no spaces. 

In [23]:
pattern = r"#[A-z0-9]+"
result = re.search(pattern, tw)
result

<re.Match object; span=(198, 211), match='#statstwitter'>

We can alternatively retrieve all the matches, while throwing away the positional information, using `re.findall()`: 

In [24]:
re.findall(pattern, tw)

['#statstwitter',
 '#epitwitter',
 '#rstats',
 '#math',
 '#AI',
 '#DataScience',
 '#python',
 '#Science']

We achieved our goal! In the next lecture, we'll look at how to extract even more complex expressions. 